velox/benchmarks/basic/LikeBenchmark.cpp

```diff
@@ -0,0 +1,86 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/Benchmark.h>
+#include <folly/init/Init.h>
+
+#include "velox/benchmarks/ExpressionBenchmarkBuilder.h"
+#include "velox/functions/lib/Re2Functions.h"
+#include "velox/functions/prestosql/registration/RegistrationFunctions.h"
+
+using namespace facebook;
+using namespace facebook::velox;
+using namespace facebook::velox::functions;
+using namespace facebook::velox::functions::test;
+using namespace facebook::velox::memory;
+using namespace facebook::velox;
+
+int main(int argc, char** argv) {
+  folly::Init init(&argc, &argv);
+
+  exec::registerStatefulVectorFunction("like", likeSignatures(), makeLike);
+  // Register the string functions we need.
+  prestosql::registerStringFunctions("");
+
+  // exec::register
+  ExpressionBenchmarkBuilder benchmarkBuilder;
+  const vector_size_t vectorSize = 1000;
+  auto vectorMaker = benchmarkBuilder.vectorMaker();
+
+  auto makeInput =
+      [&](vector_size_t vectorSize, bool padAtHead, bool padAtTail) {
+        return vectorMaker.flatVector<std::string>(vectorSize, [&](auto row) {
+          // Strings in even rows contain/start with/end with a_b_c depends on
+          // value of padAtHead && padAtTail.
+          if (row % 2 == 0) {
+            auto padding = std::string("x", row / 2 + 1);
+            if (padAtHead && padAtTail) {
+              return fmt::format("{}a_b_c{}", padding, padding);
+            } else if (padAtHead) {
+              return fmt::format("{}a_b_c", padding);
+            } else if (padAtTail) {
+              return fmt::format("a_b_c{}", padding);
+            } else {
+              return std::string("a_b_c");
+            }
+          } else {
+            return std::string("x", row);
+          }
+        });
+      };
+
+  auto substringInput = makeInput(vectorSize, true, true);
+  auto prefixInput = makeInput(vectorSize, false, true);
+  auto suffixInput = makeInput(vectorSize, true, false);
+
+  benchmarkBuilder
+      .addBenchmarkSet(
+          "like",
+          vectorMaker.rowVector(
+              {"col0", "col1", "col2"},
+              {substringInput, prefixInput, suffixInput}))
+      .addExpression("like_substring", R"(like(col0, '%a\_b\_c%', '\'))")
+      .addExpression("like_prefix", R"(like(col1, 'a\_b\_c%', '\'))")
+      .addExpression("like_suffix", R"(like(col2, '%a\_b\_c', '\'))")
+      .addExpression("like_generic", R"(like(col0, '%a%b%c'))")
+      .addExpression("strpos", R"(strpos(col0, 'a_b_c'))")
+      .addExpression("starts_with", R"(starts_with(col1, 'a_b_c'))")
```

Comment:
I wonder how does this work? Don't we need to split these benchmarks into 3 different sets? 

```
// This class represents a set of expressions to benchmark those expressions
// uses the same input vector type, and are expected to have the same result
// if testing is not disabled for the set.
```

Reply:
`CastBenchmark.cpp` has similar usage? is the comment you refered to is out of date? (Have not look into the implementation details of `ExpressionBenchmarkSet` yet.)

```
  benchmarkBuilder
      .addBenchmarkSet(
          "cast",
          vectorMaker.rowVector(
              {"valid",
               "empty",
               "nan",
               "decimal",
               "short_decimal",
               "long_decimal",
               "timestamp"},
              {validInput,
               invalidInput,
               nanInput,
               decimalInput,
               shortDecimalInput,
               longDecimalInput,
               timestampInput}))
      .addExpression("try_cast_invalid_empty_input", "try_cast (empty as int) ")
      .addExpression(
          "tryexpr_cast_invalid_empty_input", "try (cast (empty as int))")
      .addExpression("try_cast_invalid_nan", "try_cast (nan as int)")
      .addExpression("tryexpr_cast_invalid_nan", "try (cast (nan as int))")
      .addExpression("try_cast_valid", "try_cast (valid as int)")
      .addExpression("tryexpr_cast_valid", "try (cast (valid as int))")
      .addExpression("cast_valid", "cast(valid as int)")
      .addExpression(
          "cast_decimal_to_inline_string", "cast (decimal as varchar)")
      .addExpression("cast_short_decimal", "cast (short_decimal as varchar)")
      .addExpression("cast_long_decimal", "cast (long_decimal as varchar)")
      .addExpression("cast_timestamp", "cast (timestamp as varchar)")
```

Reply:
CastBenchmark has .disableTesting();

CC: @laithsakka

Reply:
@mbasmanova After digging into the details of ExpressionBenchmarkBuilder, I find my understanding of it was not quite right, updated the benchmark.